### PR TITLE
Apply Prev Epoch Timeouts To Leader Index Computation

### PR DESCRIPTION
**Pull request by tholonious**
Apply Prev Epoch Timeouts To Leader Index Computation

Adjust processOrphanBlockPoS

### CODE REVIEW

**Comment by tholonious** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1172?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1172** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1172?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1171** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1171?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feature/proof-of-stake`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tholonious and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/pos_blockchain.go [hunk 1]
```diff
@@ -1169,12 +1134,12 @@ func (bav *UtxoView) hasValidProposerPartialSignaturePoS(block *MsgDeSoBlock, sn
 // block height + view number pair. It returns a bool indicating whether
 // we confirmed that the leader is valid. If we receive an error, we are unsure
 // if the leader is invalid or not, so we return false.
-func (bav *UtxoView) hasValidBlockProposerPoS(block *MsgDeSoBlock) (_isValidBlockProposer bool, _err error) {
-	currentEpochEntry, err := bav.GetCurrentEpochEntry()
+func (bc *Blockchain) hasValidBlockProposerPoS(block *MsgDeSoBlock, utxoView *UtxoView) (_isValidBlockProposer bool, _err error) {
```
Comment (lazynina) [inline lib/pos_blockchain.go:1137]: Should we rename `utxoView` to `parentUtxoView` here?

File: lib/pos_blockchain.go [hunk 2]
```diff
@@ -1192,34 +1157,52 @@ func (bav *UtxoView) hasValidBlockProposerPoS(block *MsgDeSoBlock) (_isValidBloc
...
-	// and the current block is at height 15 and view 17, then the number of timeouts that have occurred is
-	// (17 - 11) - (15 - 10) = 1. This means this block should be proposed by the 2nd leader in the schedule,
-	// which is at index 1.
-	leaderIdxUint64 := (viewDiff - heightDiff) % uint64(len(leaders))
+	// - [(block.View - currentEpoch.InitialView) - (block.Height - currentEpoch.InitialHeight) + numTimeoutsBeforeEpochTransition] % len(leaders)
+	// - The number of views that have elapsed since the start of the epoch is block.View - currentEpoch.InitialView.
+	// - The number of blocks that have been added to the chain since the start of the epoch is
+	//   block.Height - currentEpoch.InitialHeight.
+	// - The difference between the above two numbers is the number of timeouts that have occurred in this epoch.
+	// - The numTimeoutsBeforeEpochTransition is the number of epochs that have occurred during the epoch transition
```
Comment (lazynina) [inline lib/pos_blockchain.go:1175]:
```suggestion
	// - The numTimeoutsBeforeEpochTransition is the number of timeouts that have occurred during the epoch transition
```

**Review by lazynina — APPROVED** [review]

File: lib/pos_blockchain.go (hunk 1 above)
Comment (tholonious) [reply lib/pos_blockchain.go:1137]: Ah yes. That is a more representative name

**Comment by tholonious** [comment]
### Merge activity

* **Apr 3, 5:21 PM EDT**: @tholonious started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/1172).

**Pull request merged**